wandb/sdk/launch/builder/kaniko.py

```diff
@@ -0,0 +1,349 @@
+            context_tgz.add(context_path, arcname=".")
+        context_file.close()
+
+        if self.builder_config.get("cloud-provider") == "AWS":
+            boto3 = get_module("boto3", "aws requires boto3")
+            botocore = get_module("botocore", "aws requires botocore")
+
+            s3_client = boto3.client("s3")
+
+            try:
+                s3_client.upload_file(
+                    context_file.name, self.build_context_store, f"{run_id}.tgz"
+                )
+                os.remove(context_file.name)
+            except botocore.exceptions.ClientError as e:
+                os.remove(context_file.name)
+                raise LaunchError(f"Failed to upload build context to S3: {e}")
+            return f"s3://{self.build_context_store}/{run_id}.tgz"
+        # TODO: support gcp and azure cloud providers
+        elif self.builder_config.get("cloud-provider") == "gcp":
+            storage = get_module(
+                "google.cloud.storage", "gcp requires google-cloud-storage"
+            )
+
+            storage_client = storage.Client()
+            try:
+                bucket = storage_client.bucket(self.build_context_store)
+                blob = bucket.blob(f"{run_id}.tgz")
+                blob.upload_from_filename(context_file.name)
+                os.remove(context_file.name)
+            except Exception as e:
+                os.remove(context_file.name)
+                raise LaunchError(f"Failed to upload build context to GCP: {e}")
+            return f"gs://{self.build_context_store}/{run_id}.tgz"
+        else:
+            raise LaunchError("Unsupported storage provider")
+
+    def build_image(
+        self,
+        launch_project: LaunchProject,
+        repository: Optional[str],
+        entrypoint: Optional[EntryPoint],
+        docker_args: Dict[str, Any],
+    ) -> str:
+        if repository is None:
+            raise LaunchError("repository is required for kaniko builder")
+        image_uri = f"{repository}:{launch_project.run_id}"
+        entry_cmd = get_entry_point_command(entrypoint, launch_project.override_args)
+
+        # kaniko builder doesn't seem to work with a custom user id, need more investigation
+        dockerfile_str = generate_dockerfile(
+            launch_project, entry_cmd, launch_project.resource, self.type
+        )
+        create_metadata_file(
+            launch_project,
+            image_uri,
+            sanitize_wandb_api_key(entry_cmd),
+            docker_args,
+            sanitize_wandb_api_key(dockerfile_str),
+        )
+        context_path = _create_build_ctx(launch_project, dockerfile_str)
+        run_id = launch_project.run_id
+
+        _, api_client = get_kube_context_and_api_client(
+            kubernetes, launch_project.resource_args
+        )
+        build_job_name = f"{self.build_job_name}-{run_id}"
+        config_map_name = f"{self.config_map_name}-{run_id}"
```

Comment:
Won't this config map be the same for every run_id?  It might be better to just create the global config map for our setup if it doesn't exist and always use that.  Seems a bit silly to create a separate one for each run.

Reply:
Different runs may have a different generated dockerfiles which will lead to different values stored in the configmap

Reply:
I guess I was maybe confusing this with the ECR config map that tells kaniko how to authenticate with the registry.  That should stay the same for all builds in a given namespace I would think...

Reply:
Thats a good point, moving forward I'm thinking that we'll have a single builder created when a launch agent is instantiated. The builder could create the ecr config map then. Right now, since a new builder is instantiated every time a new job is received we would need to move this into the agent set up and tear down.

I can move this into a function that runs during agent set up, and delete it on teardown to make it cleaner, but in the future this should be put back into the builder.

Reply:
No need to refactor for now, just something I noticed we should track for a future revision.